lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -2091,6 +2091,245 @@ auto AssociatedTypeInference::solve(ConformanceChecker &checker)
   return None;
 }
 
+TypeWitnessSystem::TypeWitnessSystem(
+    ArrayRef<AssociatedTypeDecl *> assocTypes) {
+  for (auto *assocType : assocTypes) {
+    this->TypeWitnesses.try_emplace(assocType->getName());
+  }
+}
+
+TypeWitnessSystem::~TypeWitnessSystem() {
+  for (auto *equivClass : this->EquivalenceClasses) {
+    delete equivClass;
+  }
+}
+
+bool TypeWitnessSystem::hasResolvedTypeWitness(Identifier name) const {
+  return (bool)getResolvedTypeWitness(name);
+}
+
+Type TypeWitnessSystem::getResolvedTypeWitness(Identifier name) const {
+  assert(this->TypeWitnesses.count(name));
+
+  if (auto *equivClass = this->TypeWitnesses.lookup(name).EquivClass) {
+    return equivClass->ResolvedTy;
+  }
+
+  return Type();
+}
+
+const AssociatedTypeDecl *
+TypeWitnessSystem::getDefaultedAssocType(Identifier name) const {
+  assert(this->TypeWitnesses.count(name));
+
+  return this->TypeWitnesses.lookup(name).DefaultedAssocType;
+}
+
+void TypeWitnessSystem::addTypeWitness(Identifier name, Type type) {
+  assert(this->TypeWitnesses.count(name));
+
+  if (const auto *depTy = type->getAs<DependentMemberType>()) {
+    // If the type corresponds to a name variable in the system, form an
+    // equivalence between variables.
+    if (depTy->getBase()->is<GenericTypeParamType>()) {
+      if (this->TypeWitnesses.count(depTy->getName())) {
+        return addEquivalence(name, depTy->getName());
+      }
+    } else {
+      while (depTy->getBase()->is<DependentMemberType>()) {
+        depTy = depTy->getBase()->castTo<DependentMemberType>();
+      }
+
+      // Equivalences of the form 'Self.X == Self.X.*' do not contribute
+      // to solving the system, so just ignore them.
```

Comment:
What about Self.X.Y == Self.X.Y.*, etc?

Reply:
Anything other than `Self.X := Y` is not considered at all.